### DISCUSSION TITLE
Offset problem with "keep_with_next (row)" option in a table.

### DISCUSSION THREAD

**Discussion by PhilippePerret** [prawnpdf/prawn-table#153] [Q&A] [answered]
Hello, 

I hope there are still people passing by here... 

I would like to use the "keep with next" option to make one (or more) row stay "attached" to another even when the other one moves to the next page.

To do this, in the cell options, I implemented the "keep_with_next" option, which can be set to "true" or the number of rows to keep together. 

~~~ruby
module PrawnTableCellTextExtension
  def keep_with_next=(value)
    value = nil if value === false
    @_keep_with_next = value
  end
  def keep_with_next?
    not(@_keep_with_next.nil?)
  end
  def keep_with_next
    if @_keep_with_next === true
      1
    else 
      @_keep_with_next
    end
  end
end

module Prawn
  class Table
    class Cell
      class Text
        include PrawnTableCellTextExtension
      end
    end
  end
end
~~~

I wrapped the "start_new_page?" method for additional processing. 

~~~ruby
module Prawn
  class Table
    
    alias :original_start_new_page? :start_new_page?
    
    def start_new_page?(cell, offset, ref_bounds)
      if cell.respond_to?(:keep_with_next?) # so not a dummy
        if cell.keep_with_next?
          # 
          # As much as rows we want
          # 
          extra_h = 0
          cell.keep_with_next.times do |i|
            extra_h += cells.rows(cell.row + i).height
          end
          offset -= extra_h + 5 #  <===== "+5"?
          # doesn't work (for one specific row) without +5
        end
      end
      original_start_new_page?(cell, offset, ref_bounds)
    end
end
~~~

Unfortunately, as you can see, I have to add the value 5 for it to work, without understanding why. 

I hope the code I provide below will help you understand. If not, I can, of course, provide more context. 

Thank you in advance for shedding light on this.

Cheers

**Comment by PhilippePerret** [comment]
Upon reviewing my code, I believe I have understood... Unfortunately, I cannot verify it yet.

That is wrong:

~~~ruby 
cell.keep_with_next.times do |i|
  extra_h += cells.rows(cell.row + i).height
end
~~~

This might be better:

~~~ruby
(cell.keep_with_next + 1).times do |i|
  extra_h += cells.rows(cell.row + i).height
end
~~~

**Reply by pointlessone** [reply]
1. 5 seems like a small extra offset. Is it 5 for any number of rows to stick together or is it different for differently sized batches?
2. In your revised code `+ 1` should probably go not into the first line but into the second where you look up the consequent rows. With `times` it starts from 0. So your version would count the current and the n next rows for `extra_h`. You probably want only the next n rows.

**Reply by PhilippePerret** [reply, marked as answer]
Thanks @pointlessone (and congratulations for your awesome work on Prawn/table by the way).

The final workable code is:

~~~ruby
module Prawn
  class Table
    
    alias :original_start_new_page? :start_new_page?
    
    def start_new_page?(cell, offset, ref_bounds)
      if cell.respond_to?(:keep_with_next?) # so not a dummy
        if cell.keep_with_next?
          # 
          # As much as rows we want
          # 
          (cell.keep_with_next + 1).times do |i|
            offset -= cells.rows(cell.row + i).height
          end
        end
      end
      original_start_new_page?(cell, offset, ref_bounds)
    end
end
~~~
